apps/src/templates/teacherDashboard/AddSectionDialog.jsx

```diff
@@ -0,0 +1,159 @@
+  extras: 'yes',
+  pairing: 'yes',
+};
+
+export class AddSectionDialog extends Component {
+  static propTypes = {
+    handleClose: PropTypes.func,
+    isOpen: PropTypes.bool,
+    // Provided by Redux
+    updateSection: PropTypes.func.isRequired,
+  };
+
+  state = {
+    ...initialState,
+  };
+
+  handleClose = () => {
+    this.props.handleClose();
+    this.setState(initialState);
+  };
+
+  handleLoginChoice = (loginType) => {
+    this.setState({loginType});
+  };
+
+  handleNameChange = (name) => {
+    this.setState({name});
+  };
+
+  handleGradeChange = (grade) => {
+    this.setState({grade});
+  };
+
+  handleExtrasChange = (extras) => {
+    this.setState({extras});
+  };
+
+  handlePairingChange = (pairing) => {
+    this.setState({pairing});
+  }
+
+  onClickEditSave = () => {
+    const {updateSection} = this.props;
+    //const persistedSection = false;
+
+    const selectedAssignment = this.assignment.getSelectedAssignment();
+    const data = {
+      //id: null,
+      name: this.state.name,
+      login_type: this.state.loginType,
+      grade: this.state.grade,
+      stage_extras: this.state.extras === 'yes' ? true : false,
+      pairing_allowed: this.state.pairing === 'yes' ? true : false,
+      course_id: selectedAssignment ? selectedAssignment.courseId : null,
+    };
+
+    if (selectedAssignment && selectedAssignment.scriptId) {
+      data.script = {
+        id: selectedAssignment.scriptId
+      };
+    }
+
+    const suffix = '';
+    const sectionId = -1; // When it's a new section
+
+    $.ajax({
+      url: `/v2/sections${suffix}`,
+      method: 'POST',
+      contentType: 'application/json;charset=UTF-8',
+      data: JSON.stringify(data),
+    }).done(result => {
+      updateSection(sectionId, result);
+      // close modal after save
+      this.handleClose();
+    }).fail((jqXhr, status) => {
+      // We may want to handle this more cleanly in the future, but for now this
+      // matches the experience we got in angular
+      alert(i18n.unexpectedError());
+      console.error(status);
+    });
+  };
+
+  renderContent() {
+    const title = i18n.newSection();
+    if (!this.state.loginType) {
+      return (
+        <AddInitialStudentsView
+          title={title}
+          handleLoginChoice={this.handleLoginChoice}
+          handleCancel={this.handleClose}
+        />
+      );
+    } else {
+      return (
+        <EditSectionForm
+          title={title}
+          assignmentRef = {(element) => this.assignment = element}
+          handleSave={this.onClickEditSave}
+          handleClose={this.handleClose}
+          name={this.state.name}
+          handleName={this.handleNameChange}
+          grade={this.state.grade}
+          handleGrade={this.handleGradeChange}
+          extras={this.state.extras}
+          handleExtras={this.handleExtrasChange}
+          pairing={this.state.pairing}
+          handlePairing={this.handlePairingChange}
+        />
+      );
+    }
+  }
```

Comment:
At the risk of rabbitholing:

The `Switch/Case` thing is verbose, but I still like it because I feel like I hit this conundrum all the time.  It's definitely more general than the ternary case though.

I don't actually like the `<Ternary>` you presented though because I'm not sure that in practice it will always be obvious that `A` is the true case and `B` is the else case (if the ternary gets long) and I'm not sure what the expected behavior is if we don't pass exactly two children.

I'd almost prefer they got passed as props:

```jsx
<Ternary
  if={condition}
  then={<A/>}
  else={<B/>}
/>
```
but that's also not going to scale up nicely when A and B have their own props.